### ISSUE TITLE
[Bug]: SM80 router logits round through BF16 for M>8; backport upstream #54048

### ISSUE THREAD

**Issue by Morrowmake** [wtdcode/vllm-backport#65]
## Summary
Please consider backporting the merged upstream router fix [vllm-project/vllm#54048](https://github.com/vllm-project/vllm/pull/54048) (merge `b5707bf994cb968adfc7a29fbb80b0582f53f38d`). The current fork head `85f227da62bbcc851cdf32a841b4b92017a0f02f` still restricts the BF16-input/FP32-output cuBLAS tier to newer architectures.

On SM80, identical synthetic token rows give exact FP32 logits at M=1/8, but the M=9/33 fallback rounds logits through BF16 and changes the selected expert set. This is a numerical dispatch issue, not ordinary FP32 reduction-order variation: each relevant dot contains only two exact dyadic products.

## Environment
- Backport source: `85f227da62bbcc851cdf32a841b4b92017a0f02f`
- Image: `lazymio/vllm-backport@sha256:24e5d7c07d61ec50abbe36c1c755cf1310dd9bec0de1d1bf7c777442a1e3733b`
- NVIDIA CMP 170HX, compute capability 8.0; PyTorch `2.13.0+cu130`; TF32 disabled in this reproduction.
- No checkpoint/model loading or private input data is required.

## Expected and observed
The exact logit for expert7 is `1 + 1/512 = 1.001953125`. Expert8 has logit1 and a selection-only correction of `1/8192`. Correct FP32 logits select experts0..7. BF16 output rounding erases expert7's advantage and selects0..6,8.

The original fork passes the exact logit/expert checks at M=1/8 and fails at M=9/33. Applying only the upstream architecture-independent cuBLAS eligibility change makes all four sizes pass **while retaining BF16 weight storage**. The identical script below was executed in separate offline stock/fixed processes with a 300-second guard each; both completed normally with verified teardown.

The final upstream diff covers all CUDA architectures; the PR's original prose saying Ampere was unchanged is stale.

| Router implementation | M | Expert7 logit | Selected experts | All logits exact |
|---|---|---:|---|---|
| Stock | 1, 8 | 1.001953125 | 0,1,2,3,4,5,6,7 | yes |
| Stock | 9, 33 | 1.0 | 0,1,2,3,4,5,6,8 | no |
| Upstream cuBLAS eligibility backport | 1, 8, 9, 33 | 1.001953125 | 0,1,2,3,4,5,6,7 | yes |

Reproducer SHA256: `fd609e571cb1323caf2c2a49c594c3aa49d6515b775e298ed0f4045ba40a7665`. Save the code as `upstream-repro-router.py` and run `OMP_NUM_THREADS=2 MKL_NUM_THREADS=2 OPENBLAS_NUM_THREADS=2 MAX_JOBS=2 python3 upstream-repro-router.py` in the stated backport environment.

## Minimal reproducer
```python
"""Synthetic reproduction for wtdcode/vllm-backport SM80 router rounding.
No checkpoint, network access, or model load. Tested with PyTorch 2.13.0+cu130.
"""
import json
import tempfile
from pathlib import Path

import torch
from vllm.config import VllmConfig, set_current_vllm_config
from vllm.distributed import (
    init_distributed_environment, initialize_model_parallel,
    destroy_model_parallel, destroy_distributed_environment,
)
from vllm.model_executor.layers.fused_moe.router.gate_linear import GateLinear
from vllm.model_executor.layers.fused_moe.router.grouped_topk_router import grouped_topk
from vllm.utils.torch_utils import set_default_torch_dtype


def main():
    torch.cuda.set_device(0)
    torch.backends.cuda.matmul.allow_tf32 = False
    rows = []
    with tempfile.TemporaryDirectory() as directory:
        with set_current_vllm_config(VllmConfig()), set_default_torch_dtype(torch.bfloat16), torch.inference_mode():
            init_distributed_environment(world_size=1, rank=0, local_rank=0, backend='gloo',
                distributed_init_method='file://' + str(Path(directory) / 'rendezvous'))
            initialize_model_parallel(1, 1, backend='gloo')
            try:
                with torch.device('cuda'):
                    gate = GateLinear(4096, 288, bias=False, out_dtype=torch.float32,
                                      params_dtype=torch.bfloat16)
                    gate.weight.zero_()
                    gate.weight[:7, 0] = 2
                    gate.weight[7, 0] = 1
                    gate.weight[7, 1] = 1 / 512
                    gate.weight[8, 0] = 1
                    correction = torch.zeros(288, dtype=torch.float32)
                    correction[8] = 1 / 8192
                for m in (1, 8, 9, 33):
                    x = torch.zeros((m, 4096), dtype=torch.bfloat16, device='cuda')
                    x[:, :2] = 1
                    logits, _ = gate(x)
                    _, ids = grouped_topk(x, logits, 8, True, 1, 1, 'sigmoid', 2.5, correction)
                    expected = torch.zeros((m, 288), dtype=torch.float32)
                    expected[:, :7] = 2
                    expected[:, 7] = 1 + 1 / 512
                    expected[:, 8] = 1
                    rows.append({'M': m, 'weight_dtype': str(gate.weight.dtype),
                                 'output_dtype': str(logits.dtype),
                                 'expert7_logit': float(logits[0, 7]),
                                 'selected_experts': ids[0].sort().values.cpu().tolist(),
                                 'logits_exact': torch.equal(logits.cpu(), expected)})
            finally:
                destroy_model_parallel()
                destroy_distributed_environment()
    print(json.dumps({'torch': torch.__version__, 'device': torch.cuda.get_device_name(),
                      'capability': torch.cuda.get_device_capability(), 'results': rows}, indent=2))


if __name__ == '__main__':
    main()
```

## Scope
This verifies the router operation and expert selection, not full-model quality or performance. The exact upstream dispatch change was tested on the pinned backport. No driver, firmware, power or clock changes are part of the fix. This was an AI-assisted investigation with executed native reproducers.

**Comment by wtdcode** [comment]
This looks good to me. Thanks, will have a fix.
